internal/sync2/filepipe.go

```diff
@@ -0,0 +1,275 @@
+// Copyright (C) 2018 Storj Labs, Inc.
+// See LICENSE for copying information
+
+package sync2
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"math"
+	"os"
+	"sync"
+	"sync/atomic"
+)
+
+// filepipe is a file backed pipe
+type filepipe struct {
+	file     ReadWriteCloserAt
+	refcount int32
+
+	mu     sync.Mutex
+	nodata sync.Cond
+	read   int64
+	write  int64
+	limit  int64
+
+	writerDone bool
+	writerErr  error
+
+	readerDone bool
+	readerErr  error
+}
+
+// NewFilePipe returns a pipe that uses file-system to offload memory
+func NewFilePipe(tempdir string) (PipeReader, PipeWriter, error) {
+	tempfile, err := ioutil.TempFile(tempdir, "filepipe")
+	if err != nil {
+		return nil, nil, err
+	}
+
+	pipe := &filepipe{
+		file:     tempfile,
+		refcount: 2,
+		limit:    math.MaxInt64,
+	}
+	pipe.nodata.L = &pipe.mu
+
+	return filePipeReader{pipe}, filePipeWriter{pipe}, nil
+}
+
+type filePipeReader struct{ pipe *filepipe }
+type filePipeWriter struct{ pipe *filepipe }
+
+// Close implements io.Reader Close
+func (reader filePipeReader) Close() error { return reader.CloseWithError(io.EOF) }
+
+// Close implements io.Writer Close
+func (writer filePipeWriter) Close() error { return writer.CloseWithError(io.EOF) }
```

Comment:
Shouldn't we close the writer with `io.ErrClosedPipe` instead of `io.EOF`? The latter is supposed to be returned only be Read operations.

Reply:
Yes